### PR TITLE
Try: Sticky header and pagination on Patterns page

### LINKED ISSUE #52431: Site Editor: Make the search box on the pattern page canvas sticky
## What problem does this address?
On the current Patterns page, the search box is hidden when the canvas is scrolled down.

## What is your proposed solution?
How about making this search box sticky, like the header of a All Template or All Template Parts page?

https://github.com/WordPress/gutenberg/assets/54422211/a2ed52aa-bb39-45f5-b7ca-29ac89c7e32d

### PR DESCRIPTION
**Pull request by t-hamano**
Closes: #52431

## What?
This PR sticks the header (pattern category title, pattern category description, search box, and sync statues) on the Patterns page of the Site Editor so that they are not affected by scrolling.

_edit:_ Additionally, I fixed the pagination as well.

## Why?

In the Site Editor's left sidebar navigation and template (parts) list pages, the header and title sections are fixed and are not affected by scrolling. Similarly, I believe that fixing the header on the pattern list page will make it easier to filter patterns and switch between synced categories.

I would appreciate your input on whether this change makes sense.

## How?

This PR makes no changes to the logic of the component. Just added some wrap elements to achieve Sticky Header and added some styles.

_edit:_ To fix pagination, I also split and moved components.

### Desktop view

https://github.com/WordPress/gutenberg/assets/54422211/e17aa954-a677-4465-9c8e-cea5b282f5b2

### Mobile view

https://github.com/WordPress/gutenberg/assets/54422211/3aa84478-b875-4be0-966f-764f434c9477

### CODE REVIEW

**Comment by jameskoster** [comment]
I was thinking about putting a similar PR together myself ❤️

How do you feel about making the pagination controls sticky too?

**Comment by t-hamano** [comment]
@jameskoster 

Thanks for the advice! I had forgotten about the pagination 😅 Updated PR and fixed pagination.

https://github.com/WordPress/gutenberg/assets/54422211/21d5a601-c81f-4fda-81b9-50e0b1a2e3ce

**Comment by jameskoster** [comment]
Nice :)

For the pagination, do you think we should align the contents with the save button?

Screenshot 2023-07-17 at 15 04 19

**Comment by t-hamano** [comment]
> For the pagination, do you think we should align the contents with the save button?

Makes sense! I have updated. Note that I had to change the pagination button size from `32px` to `36px` to perfectly match the height.

pagination_height

File: packages/edit-site/src/components/page-patterns/style.scss [hunk 1]
```diff
@@ -1,8 +1,19 @@
 .edit-site-patterns {
-	border: 1px solid $gray-800;
+	border-left: 1px solid $gray-800;
```
Comment (t-hamano) [inline packages/edit-site/src/components/page-patterns/style.scss:2]: I could not get the save button and pagination height to match exactly when there is a 1px border across the entire pattern canvas. Therefore, I applied the border only to the left side.

File: packages/edit-site/src/components/page-patterns/style.scss [hunk 2]
```diff
@@ -90,11 +122,8 @@
 	display: grid;
 	grid-template-columns: 1fr;
 	gap: $grid-unit-40;
-	// Small top padding required to avoid cutting off the visible outline
-	// when hovering items.
-	padding-top: $border-width-focus-fallback;
```
Comment (t-hamano) [inline packages/edit-site/src/components/page-patterns/style.scss:95]: In this PR, the pattern list has enough padding so this small padding should be unnecessary.

**Comment by jameskoster** [comment]
> Note that I had to change the pagination button size from 32px to 36px to perfectly match the height.

Sorry for that distraction, but buttons should all be either 32px or 40px, the save button needs updating (separately). I think we can still line everything up by using the right padding values, or `height` if necessary. What do you think?

**Comment by t-hamano** [comment]
> but buttons should all be either 32px or 40px, the save button needs updating (separately). 

I see. In principle, that means it should be a multiple of 8, right? In order to match the current save button height, it was necessary to use math.div, but in [40fa2db](https://github.com/WordPress/gutenberg/pull/52663/commits/40fa2db7bb1c163d6819b258b5fc0563a62cbde9), the size of the pagination button was returned to 32px.

capture

**Comment by jameskoster** [comment]
@t-hamano apologies for the detour here. Since the save area dimensions are 'incorrect' due to the 36px button, we shouldn't align the pagination with that afterall. That's probably a detail to tackle separately – fix the save button height, and adjust the pagination spacing. 

For now we can probably revert the last couple of commits and go back to the design you had here (https://github.com/WordPress/gutenberg/pull/52663#issuecomment-1638119740). Then I think we're good to get more reviews. Apologies again!

**Comment by t-hamano** [comment]
@jameskoster

Thank you for your detailed explanation. So in this PR, is it okay if we don't force the height to fit the incorrect save button area?

**Comment by jameskoster** [comment]
> So in this PR, is it okay if we don't force the height to fit the incorrect save button area?

Yes that's correct. We can do that in a follow-up PR :)

**Comment by t-hamano** [comment]
Updated!

Currently, the heights are:

- Save button Area: height **85px** (top border 1px + top padding 24px + button height **36px** + bottom padding 24px)
- Pagination Area: height **81px** ( top border 1px + top padding 24px + button height **32px** + bottom padding 24px)

Whether the height of the save button changes to 32px or 40px, I think we can adjust the height with padding.

pagination-height

**Review by glendaviesnz — APPROVED** [review]
This tested well for me - this is a great change, thanks for implementing

**Review by aaronrobertshaw — APPROVED** [review]
I was looking forward to seeing this polish being added to the Patterns page. Nice work @t-hamano ✨ 

Tests well for me also.

> Whether the height of the save button changes to 32px or 40px, I think we can adjust the height with padding.

Is it worth making this tweak now, so until the save button height is updated things align? The misalignment does stick out a little awkwardly.

**Comment by jameskoster** [comment]
> Is it worth making this tweak now, so until the save button height is updated things align?

I think it's probably okay to proceed, if we follow-up quickly to address the alignment in a new PR. Do you have bandwidth to work on that @t-hamano? If not I can probably take a look.

**Comment by t-hamano** [comment]
Thank you all for your reviews and feedback!

@jameskoster 

> I think it's probably okay to proceed, if we follow-up quickly to address the alignment in a new PR. Do you have bandwidth to work on that @t-hamano? If not I can probably take a look.

Yes, I will merge this PR as I will submit a follow-up PR soon.

**Pull request merged**